### ISSUE TITLE
[BUG] After 5 validated clips in (en) contributor gets an award every time

### ISSUE THREAD

**Issue by robovoice1** [common-voice/common-voice#4187]
**Describe the bug**

Please watch the recording:
I validated 5 sentences in english and got an award.
(I had already achieved my weekly goal with award before!)
After validating another 5 sentences i got another award ?!
My goals are set in 2 languages: (de) and (en) 105 clips record and listen per week

**To Reproduce**
1. Go to '...' https://commonvoice.mozilla.org/en/listen
2. Click on '....' validate 5 clips 
3. Get an award after every 5 clips
4.  The given awards do not match the set personal goals!

**Expected behavior**
The awards are given after 105 clips  are recorded and validated (once weekly )

**Screenrecording**

https://github.com/common-voice/common-voice/assets/92784377/9538156d-2d9f-44cf-a909-ca7a1b0072cb

**Mobile (please complete the following information):**
 - OS: [e.g. iOS] Android13 GrapheneOS
 - Browser [e.g. chrome, safari] Vanadium 
 - Version [e.g. 22] 117.0.5938.60

 **Additional Hardware (were you using headphones, an external speaker or an external microphone?):**
 - Type:
 - Model:

**Additional context**

Before i recorded i got 4 awards in english section !
Since today i get this strange behavior.

**Comment by robovoice1** [comment]
Here are more awards after only validating in (en) language
![Screenshot_20230922-221729](https://github.com/common-voice/common-voice/assets/92784377/0955f197-3657-4fe0-9e6b-29b34c5b0472)

**Comment by moz-kathyreid** [comment]
Replicated: I was able to replicate this bug by doing the following: 

1. Set a Custom Goal in English to Listen to 15 clips / day 
2. Listen to 15 clips -> award granted 
3. Listen to 5 more clips -> award granted 
4. Listen to 5 more clips -> award granted 

That is, the award is _re_-granted every time a set of 5 clips is recorded / validated. 

I replicated this on FF 138.0b2 (64-bit) for Linux, so it's not an Android / user agent issue - it's related to the platform logic. 

Investigation: digging in to this to see if it's an easy fix... 

I _think_ [this is the code that is in error](https://github.com/common-voice/common-voice/blob/d4e754c070f4c2f92243b94ecda180c56fad0133/web/src/components/pages/dashboard/awards/awards.tsx#L97C1-L119C2). 

I think there is an error with the way `.map()` is being used here that creates more awards per bucket than is correct. 

Not sure how to resolve or test this one, appears low priority so moving for now. 

Limited scope, low impact = very few Custom Goals are set in production from what I understand (always happy to be corrected).

**Comment by moz-kathyreid** [comment]
These are my notes from digging into this bug: 

* In `awards.ts`, this code block uses intervals of 1 (day) or 7 (week) to put the awards into buckets: 

```
  const buckets = awards.reduce((buckets, award) => {
    const bucket = buckets[award.days_interval];
    buckets[award.days_interval] = bucket ? [...bucket, award] : [award];
    return buckets;
  }, []);
```

* In my dev environment, I logged the value of `awards` to console: 

```
awards:  
Array [ {…}, {…} ]
​
0: Object { locale: "en", notification_seen_at: "2025-04-17T09:59:00.000Z", seen_at: "2025-04-17T09:59:04.000Z", … }
​
1: Object { locale: "en", notification_seen_at: "2025-04-17T11:52:44.000Z", seen_at: "2025-04-17T11:54:02.000Z", … }
​
length: 2
​
<prototype>: Array []
awards.tsx:97
```

* and `buckets`: 

```
buckets:  
Array [ <1 empty slot>, (2) […] ]
​
1: Array [ {…}, {…} ]
​​
0: Object { locale: "en", notification_seen_at: "2025-04-17T09:59:00.000Z", seen_at: "2025-04-17T09:59:04.000Z", … }
​​​
amount: 15
​​​
days_interval: 1
​​​
locale: "en"
​​​
notification_seen_at: "2025-04-17T09:59:00.000Z"
​​​
seen_at: "2025-04-17T09:59:04.000Z"
​​​
type: "speak"
​​​
<prototype>: Object { … }
​​
1: Object { locale: "en", notification_seen_at: "2025-04-17T11:52:44.000Z", seen_at: "2025-04-17T11:54:02.000Z", … }
​​​
amount: 15
​​​
days_interval: 1
​​​
locale: "en"
​​​
notification_seen_at: "2025-04-17T11:52:44.000Z"
​​​
seen_at: "2025-04-17T11:54:02.000Z"
​​​
type: "listen"
​​​
<prototype>: Object { … }
​​
length: 2
​​
<prototype>: Array []
​
length: 2
​
<prototype>: Array []
awards.tsx:108
```

I _suspect_ the `.reduce()` function is causing the sparse array (array with an empty slot) seen here. I also _suspect_ that this is causing some sort of division error with `DAYS_INTERVAL`. 

I wasn't able to replicate the pattern seen in production in my dev environment, even after lots of trials with multiple accounts, but I wonder if it's because my `DAYS_INTERVAL` == 1. I will test again tomorrow in which case the `DAYS_INTERVAL` will be == 2 and may present differently.

**Comment by moz-kathyreid** [comment]
I still cannot replicate this in my dev environment. It may be related to `DAYS_INTERVAL` but rather than an issue with division by 1, it may be an issue with division by 7. So I am going to put this on hold for a few days until I cross over that threshold and come back to it, rather than wasting time. 

I am **still* able to replicate this easily in the production environment - every time I validate 5 clips, I receive another Award. 

However, I want to reliably replicate in development before implementing a fix.

**Comment by moz-kathyreid** [comment]
## Notes from debugging session with @moz-dfeller 

Dmitrij kindly paired with me yesterday, and walked me through some diagnostic strategies. I am summarising both the strategies and the findings here both as a learning piece and so I have a good point from which to do more debugging.

### General strategies 

* Frontend - using `console.log` commands, understand the contents of variables that are being manipulated and used to render the page.

* Frontend - using XHR, better understand what is being served by `api.ts`. The key endpoints here are `user_client` (which contains an `award` object and also a `custom_goals` object) and `goals`, which contains data on the _standard_ goals that have been defined, and a count of clips added/reviewed, but _not_ information on _custom_ goals. 

* Backend - tracing through `api.ts` to better understand how the API functions. In particular, `user_client.ts` and `awards.ts` handle the retrieval and insertion, respectively, of Awards. 



### Findings 

* In general, we found that this behaviour is likely from multiple errors in the front-end _and_  the back-end. On the front-end, there are some logic problems with the `.reduce()` function where it doesn't check to see if a `bucket` is empty. 

* But in general, the front-end is rendering what is returned from the API correctly, and  there is likely a logic bug in the insertion of Awards into the database, possibly missing a check for whether an Award has _already_ been obtained during the time interval (daily or weekly). This _may_ be related to time zones - for example, the CV infrastructure runs on GMT / Zulu and the client will run on the data contributor's local time zone (e.g. UC + 1000hrs for me). 

Example of awards being returned from API in `user_client` (showing they have likely been added erroneously to the db): 

```
{
	"awards": [
		{
			"locale": "en",
			"notification_seen_at": "2025-04-09T10:16:07.000Z",
			"seen_at": "2025-04-09T10:16:16.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": "2025-04-09T12:10:03.000Z",
			"seen_at": "2025-04-09T10:23:37.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": "2025-04-09T12:10:03.000Z",
			"seen_at": "2025-04-09T10:27:18.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": "2025-04-09T12:10:03.000Z",
			"seen_at": "2025-04-10T10:08:52.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": "2025-04-17T09:45:55.000Z",
			"seen_at": "2025-04-17T09:46:37.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": "2025-04-17T10:02:34.000Z",
			"seen_at": "2025-04-17T09:46:37.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": "2025-04-17T10:02:34.000Z",
			"seen_at": "2025-04-17T09:47:48.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": "2025-04-17T12:15:51.000Z",
			"seen_at": "2025-04-17T12:15:56.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": "2025-04-23T09:21:31.000Z",
			"seen_at": "2025-04-23T09:22:14.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": "2025-04-23T09:35:22.000Z",
			"seen_at": "2025-04-23T09:22:14.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": "2025-04-23T09:35:22.000Z",
			"seen_at": "2025-04-23T09:24:25.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": "2025-04-23T11:21:05.000Z",
			"seen_at": "2025-04-23T11:21:18.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		},
		{
			"locale": "en",
			"notification_seen_at": null,
			"seen_at": "2025-04-23T11:22:47.000Z",
			"type": "listen",
			"days_interval": 1,
			"amount": 15
		}
	]
}
```

* `awards.ts`: 
  - The function `checkProgress()` performs insertion of the Award into the database. There may be insufficient logic here, e.g. to check if an Award _should_ be inserted or not. 

### Next steps 

* Better understand the insertion logic for `checkProgress()`
* And use this to replicate the bug in the development environment

**Comment by moz-kathyreid** [comment]
Another thought: In production, I have two languages in my Profile - `en` and `sw` - and in development I only have one. The reporter also works across two languages - `en` and `de`. This may be a distinguishing feature for why I can't replicate this in dev. 

Could multiple languages play a role here - for example checking of `LOCALE` before insertion into the database?

**Comment by moz-kathyreid** [comment]
* Ruled out the two languages theory by adding another language -> still can't replicate in development

Current hypothesis and diagnostics: 

* That there is a time zone discrepancy in the calculations around whether 
  - an Award threshold has been reached based on the `START_INTERVAL` of the custom goal 
  - and whether an Award should be granted 

For example, this code in `awards.ts` strips out the timezone information of the `current_interval_start` of the custom goal to insert it into the db. If the timezone of `goal.current_interval_start` is significantly different from the timezone of the database - e.g. many hours (like AEST / AEDT which is 10hrs / 11hrs + from UTC) then I suspect it is calculating the interval start incorrectly, allowing for a new `Award` to be granted.

```
const intervalStart = new Date(goal.current_interval_start)
      .toISOString()
      .slice(0, 19)
      .replace('T', ' ')
```

Testing this or proving this is difficult, but this is what I've tried: 

* Using an extension to spoof the timezone in my browser => this made no difference, probably because the timezone is being stripped on the current interval start, and the current interval has "already started". 

* Checking what timezone the MySQL db uses => 
`SELECT @@GLOBAL.time_zone, @@SESSION.time_zone;`
This returned `SYSTEM`, which means it's the same time zone as my system (UTC +1000hrs AEST), which is the same time zone as my browser. This explains, I think, why I can't replicate on my system - as there is no discrepancy b/w browser and the MySQL db. 

Next steps I want to try: 

* Changing the time zone of the MySQL db to be UTC / GMT but leaving my browser as UTC + 1000hrs AEST 
* Starting a _new_ custom goal, because the `intervalStart` will be stripped of timezone information, and this _should_ cause a discrepancy in the `current_interval_start` timestamp 

More broadly

* I think the date conversion of `intervalStart` is not needed, because the field `awards.goal_interval_start` is of type TIMESTAMP, _not_ DATETIME, so I think the conversion is not necessary.

**Comment by moz-kathyreid** [comment]
This is blocked on me learning more about Jest and the correct way to mock the DB. I have a Typescript reference book and plan to teach myself. Then I can better understand what the problematic SQL is doing.

**Comment by moz-bozden** [comment]
This seems to be a race condition, we SELECT and then INSERT.
Although the code tries to handle it by rounding it to seconds, somehow:
- Two requests comes, checks that there is not such award
- And inserts two awards

Correct solution would need unique indexes including time, and INSERT IGNORE ... UPDATE flow...
Indexing means downtime...
